### PR TITLE
Implement wand-like optimization for boolean query

### PR DESCRIPTION
**Pull request by MBkkt**
Also:
* Compute score single time
* Make wand score compatible work
* Add wand optimization for conjunction, disjunction and min_match

master:
```
Index read time calls:1, time: 50293 us, avg call: 50293 us
Order build time calls:1, time: 1 us, avg call: 1 us
Query building (AndHighHigh) time calls:1000, time: 27679 us, avg call: 27.679 us
Query building (AndHighLow) time calls:1000, time: 20964 us, avg call: 20.964 us
Query building (AndHighMed) time calls:1000, time: 28698 us, avg call: 28.698 us
Query building (HighTerm) time calls:282, time: 1560 us, avg call: 5.53191 us
Query building (LowTerm) time calls:1000, time: 8037 us, avg call: 8.037 us
Query building (MedTerm) time calls:1000, time: 6794 us, avg call: 6.794 us
Query building (MinMatch2High2Med) time calls:2, time: 320 us, avg call: 160 us
Query building (Or4High) time calls:2, time: 78 us, avg call: 39 us
Query building (Or6High4Med2Low) time calls:2, time: 203 us, avg call: 101.5 us
Query execution (AndHighHigh) time calls:1000, time: 84034036 us, avg call: 84034 us
Query execution (AndHighLow) time calls:1000, time: 1891038 us, avg call: 1891.04 us
Query execution (AndHighMed) time calls:1000, time: 29832592 us, avg call: 29832.6 us
Query execution (HighTerm) time calls:282, time: 55568 us, avg call: 197.05 us
Query execution (LowTerm) time calls:1000, time: 240645 us, avg call: 240.645 us
Query execution (MedTerm) time calls:1000, time: 230304 us, avg call: 230.304 us
Query execution (MinMatch2High2Med) time calls:2, time: 1095620 us, avg call: 547810 us
Query execution (Or4High) time calls:2, time: 1831226 us, avg call: 915613 us
Query execution (Or6High4Med2Low) time calls:2, time: 3189758 us, avg call: 1.59488e+06 us
Total Time calls:1, time: 384756430 us, avg call: 3.84756e+08 us
```

this branch:
```
Index read time calls:1, time: 49971 us, avg call: 49971 us
Order build time calls:1, time: 1 us, avg call: 1 us
Query building (AndHighHigh) time calls:1000, time: 21830 us, avg call: 21.83 us
Query building (AndHighLow) time calls:1000, time: 20462 us, avg call: 20.462 us
Query building (AndHighMed) time calls:1000, time: 22017 us, avg call: 22.017 us
Query building (HighTerm) time calls:282, time: 1846 us, avg call: 6.5461 us
Query building (LowTerm) time calls:1000, time: 6599 us, avg call: 6.599 us
Query building (MedTerm) time calls:1000, time: 6743 us, avg call: 6.743 us
Query building (MinMatch2High2Med) time calls:2, time: 314 us, avg call: 157 us
Query building (Or4High) time calls:2, time: 99 us, avg call: 49.5 us
Query building (Or6High4Med2Low) time calls:2, time: 208 us, avg call: 104 us
Query execution (AndHighHigh) time calls:1000, time: 32042074 us, avg call: 32042.1 us
Query execution (AndHighLow) time calls:1000, time: 1538670 us, avg call: 1538.67 us
Query execution (AndHighMed) time calls:1000, time: 15533962 us, avg call: 15534 us
Query execution (HighTerm) time calls:282, time: 55386 us, avg call: 196.404 us
Query execution (LowTerm) time calls:1000, time: 228807 us, avg call: 228.807 us
Query execution (MedTerm) time calls:1000, time: 227585 us, avg call: 227.585 us
Query execution (MinMatch2High2Med) time calls:2, time: 809330 us, avg call: 404665 us
Query execution (Or4High) time calls:2, time: 1342383 us, avg call: 671192 us
Query execution (Or6High4Med2Low) time calls:2, time: 2340679 us, avg call: 1.17034e+06 us
Total Time calls:1, time: 316445536 us, avg call: 3.16446e+08 us
```

### CODE REVIEW

File: core/search/conjunction.hpp [hunk 1]
```diff
@@ -79,182 +87,360 @@ struct score_iterator_adapter {
...
+    }
+    auto& merger = static_cast<Merger&>(*this);
+  align_leafs:
+    target = ShallowSeekImpl(target);
+  align_docs:
+    if (IRS_UNLIKELY(doc_limits::eof(target))) {
+      return Seal();
+    }
+    auto it = this->itrs_.begin();
+    const auto seek_target = (*it)->seek(target);
```
Comment (Dronplane) [inline core/search/conjunction.hpp:303]: Suggest to remove `seek_target` and just always store sekk result in `target`

Comment (MBkkt) [reply core/search/conjunction.hpp:303]: Well, it's not possible check line 315

Comment (Dronplane) [reply core/search/conjunction.hpp:303]: True.

File: core/search/conjunction.hpp [hunk 2]
```diff
@@ -79,182 +87,360 @@ struct score_iterator_adapter {
...
+    }
+    IRS_ASSERT(doc <= leafs_doc_);
+  eof_check:
+    if (IRS_UNLIKELY(doc_limits::eof(target))) {
+      return target;
+    }
+
+    auto max_leafs = doc_limits::eof();
+    auto min_leafs = doc;
+    score_t max_leafs_score = 0.f;
```
Comment (Dronplane) [inline core/search/conjunction.hpp:399]:
```suggestion
    score_t sum_leafs_score = 0.f;
```

File: core/search/conjunction.hpp [hunk 3]
```diff
@@ -79,182 +87,360 @@ struct score_iterator_adapter {
...
+      if (max_leafs < min_leaf) {
+        goto eof_check;
+      }
+      if (min_leafs < min_leaf) {
+        min_leafs = min_leaf;
+      }
+      if (max_leafs > max_leaf) {
+        max_leafs = max_leaf;
+      }
+      IRS_ASSERT(min_leafs < max_leafs);
```
Comment (Dronplane) [inline core/search/conjunction.hpp:417]:
```suggestion
      IRS_ASSERT(min_leafs <= max_leafs);
```
Is this possible?

**Review by Dronplane — COMMENTED** [review]
LGTM in general. Some minor comments.

File: core/search/disjunction.hpp [hunk 4]
```diff
@@ -1097,12 +1111,49 @@ class block_disjunction : public doc_iterator,
...
+            if constexpr (traits_type::kMinMatch) {
+              sum += (*it)->max.tail;
+              if (arg > sum) {  // TODO(MBkkt) strict wand: >=
+                ++min_match;
+                next = it + 1;
+              }
+            }
+            const auto others = self.scores_.sum_score - (*it)->max.tail;
+            if (arg > others) {
+              (*it)->Min(arg - others);
```
[the remark this answers is not in the document]
Comment (MBkkt) [reply core/search/disjunction.hpp:1138]: We couldn't because two part in single cycle.

[the remark this answers is not in the document]
Comment (Dronplane) [reply core/search/disjunction.hpp:1138]: But what is the reason to continue? We could just seal this disjunction to eof state as it clearly can not find anything to fulfil the threshol.

**Review by Dronplane — APPROVED** [review]
LGTM

**Pull request merged**